### PR TITLE
RFC: TimeLogger helper class

### PR DESCRIPTION
**Pull request by carlosgjs**
RFC PR (Request for comments!). We have several places in the code where we measure and log timings. This is an attempt at simplifying things and reducing the amount of code it takes to do this. This PR includes a helper class as well as changes to `S1_fft_cc_MPI.py` to show the usage.

TimeLogger is utility class to measure and log the time spent in code fragments. The basic usage is to call:
```
tlog = TimeLogger()
...
tlog.log(message)
```
This will measure the time between the last time checkpoint and the call to ``log()``. The
checkpoint is updated when:
 - ``TimeLogger`` instance is created
 - ``log()`` is called
 - ``reset()`` is called

Alternatively, an explicit reference start time may be passed when logging: ``log(message, start_time)``. E.g.:
```
start_time = tlog.reset()
sleep(1)
tlog.log("step 1")
sleep(2)
tlog.log("step 2")
sleep(1)
tlog.log("overall", start_time)
```
Would output:
```
2023-05-09 14:04:02,570 INFO utils log TIMING: 1.0020 for step 1
2023-05-09 14:04:04,575 INFO utils log TIMING: 2.0056 for step 2
2023-05-09 14:04:05,579 INFO utils log TIMING: 4.0114 for overall
```

### CODE REVIEW

File: tests/test_cli.sh [hunk 1]
```diff
@@ -0,0 +1,8 @@
+RUNNER_TEMP=~/test_temp
```
Comment (carlosgjs) [inline tests/test_cli.sh:1]: Unrelated but opportunistically adding this scripts that helps do e2e testing

File: src/noisepy/seis/S1_fft_cc_MPI.py [hunk 2]
```diff
@@ -70,7 +70,9 @@ def cross_correlate(
     # #########PROCESSING SECTION##########
     #######################################
 
-    tt0 = time.time()
+    tlog = TimeLogger(logger, logging.INFO)
+    # tlog.enabled = False
```
Comment (aniketfadia96) [inline src/noisepy/seis/S1_fft_cc_MPI.py:74]: Can we remove this comment line?

File: src/noisepy/seis/utils.py [hunk 3]
```diff
@@ -24,3 +27,24 @@ def fs_join(path1: str, path2: str) -> str:
...
+        self.time = time.time()
+        return self.time
+
+    def log(self, message: str = None, start: float = -1.0) -> float:
+        stop = time.time()
+        dt = stop - self.time if start <= 0 else stop - start
+        self.reset()
+        if self.enabled:
+            self.logger.log(self.level, f"TIMING: {dt:6.4f} for {message}")
+        return self.time
```
Comment (aniketfadia96) [inline src/noisepy/seis/utils.py:75]: Just curious to know - Why are we returning `self.time` for this function?

**Review by aniketfadia96 — COMMENTED** [review]
Added minor comments. LGTM!

File: src/noisepy/seis/utils.py [hunk 4]
```diff
@@ -24,3 +27,49 @@ def fs_join(path1: str, path2: str) -> str:
         return posixpath.join(path1, path2)
     else:
         return os.path.join(path1, path2)
+
+
+class TimeLogger:
```
Comment (aniketfadia96) [inline src/noisepy/seis/utils.py:32]: Just thinking out loud - Can we add this class as part of `rse_tooklit` package in the future?

File: src/noisepy/seis/S1_fft_cc_MPI.py [hunk 5]
```diff
@@ -70,7 +70,9 @@ def cross_correlate(
     # #########PROCESSING SECTION##########
```
Comment (uwcdc) [inline src/noisepy/seis/S1_fft_cc_MPI.py:70]: I would consider converting to a block comment for readability. For example:

```
"""
PROCESSING SECTION
"""
```

File: src/noisepy/seis/S1_fft_cc_MPI.py [hunk 6]
```diff
@@ -70,7 +70,9 @@ def cross_correlate(
...
     #######################################
 
-    tt0 = time.time()
+    tlog = TimeLogger(logger, logging.INFO)
+    # tlog.enabled = False
+    t_all = tlog.reset()
```
Comment (uwcdc) [inline src/noisepy/seis/S1_fft_cc_MPI.py:75]: Could we use a more descriptive variable here? `t_all` makes me think of setting a global time variable, not resetting the global log time.

File: src/noisepy/seis/S1_fft_cc_MPI.py [hunk 7]
```diff
@@ -94,18 +96,18 @@ def cross_correlate(
 
     # MPI loop: loop through each user-defined time chunk
     for ick in range(rank, splits, size):
-        t10 = time.time()
         ts = timespans[ick]
         if cc_store.is_done(ts):
             continue
 
         # ###########LOADING NOISE DATA AND DO FFT##################
```
Comment (uwcdc) [inline src/noisepy/seis/S1_fft_cc_MPI.py:103]:
```suggestion
        """
        LOADING NOISE DATA AND DO FFT
        """
```

File: src/noisepy/seis/S1_fft_cc_MPI.py [hunk 8]
```diff
@@ -94,18 +96,18 @@ def cross_correlate(
...
-        t0 = time.time()
+        t_chunk = tlog.reset()
```
Comment (uwcdc) [inline src/noisepy/seis/S1_fft_cc_MPI.py:104]: More descriptive variable?

File: src/noisepy/seis/utils.py (hunk 3 above)
Comment (carlosgjs) [reply src/noisepy/seis/utils.py:75]: It's convenient if you want to use an explicit time reference, like in the example:
```
start_time = tlog.reset()
...
tlog.log("overall", start_time)
```

File: src/noisepy/seis/utils.py (hunk 4 above)
Comment (carlosgjs) [reply src/noisepy/seis/utils.py:32]: Yes :)

**Review by uwcdc — COMMENTED** [review]
Overall LGTM

File: src/noisepy/seis/S1_fft_cc_MPI.py (hunk 8 above)
Comment (carlosgjs) [reply src/noisepy/seis/S1_fft_cc_MPI.py:104]: This is being used to measure the time to process a whole `chunk` (see line 240):
```
tlog.log(f"Process the chunk of {ts}", t_chunk)
```
Seemed descriptive to me? :). What did you have in mind?

File: src/noisepy/seis/S1_fft_cc_MPI.py (hunk 5 above)
Comment (carlosgjs) [reply src/noisepy/seis/S1_fft_cc_MPI.py:70]: Yeah, there's a lot of style and formatting work remaining in this code base, but not part of this PR

File: src/noisepy/seis/S1_fft_cc_MPI.py (hunk 8 above)
Comment (uwcdc) [reply src/noisepy/seis/S1_fft_cc_MPI.py:99]: Gotcha! That makes sense. I would just add a comment with that the first time it's used then.

Comment (carlosgjs) [reply src/noisepy/seis/S1_fft_cc_MPI.py]: Added comment

**Comment by uwcdc** [comment]
Looks awesome! Great job!

**Review by uwcdc — APPROVED** [review]
LGTM. Great job!

**Pull request merged**